File: apps/sandbox-runtime/internal/sessionprotocol/data_frame.go
package sessionprotocol

import (
	"encoding/binary"
	"fmt"
)

const (
	DataFrameHeaderByteLength  = 6
	DataFrameKindData          = 0x01
	PayloadKindRawBytes        = 0x01
	PayloadKindWebSocketText   = 0x02
	PayloadKindWebSocketBinary = 0x03
	MaxStreamID                = 0xffff_ffff
	DefaultStreamWindowBytes   = 64 * 1024
)

type StreamDataFrame struct {
	FrameKind   byte
	StreamID    uint32
	PayloadKind byte
	Payload     []byte
}

func validateStreamID(streamID uint32) error {
	if streamID == 0 || streamID > MaxStreamID {
		return fmt.Errorf("streamId must be an integer between 1 and %d", MaxStreamID)
	}

	return nil
}

func validatePayloadKind(payloadKind byte) error {
	switch payloadKind {
	case PayloadKindRawBytes, PayloadKindWebSocketText, PayloadKindWebSocketBinary:
		return nil
	default:
		return fmt.Errorf("payloadKind is not supported: %d", payloadKind)
	}
}

func EncodeDataFrame(input struct {
	StreamID    uint32
	PayloadKind byte
	Payload     []byte
}) ([]byte, error) {
	if err := validateStreamID(input.StreamID); err != nil {
		return nil, err
	}
	if err := validatePayloadKind(input.PayloadKind); err != nil {
		return nil, err
	}

	encoded := make([]byte, DataFrameHeaderByteLength+len(input.Payload))
	encoded[0] = DataFrameKindData
	binary.BigEndian.PutUint32(encoded[1:5], input.StreamID)
	encoded[5] = input.PayloadKind
	copy(encoded[DataFrameHeaderByteLength:], input.Payload)
	return encoded, nil
}

func DecodeDataFrame(encoded []byte) (StreamDataFrame, error) {
	if len(encoded) < DataFrameHeaderByteLength {
		return StreamDataFrame{}, fmt.Errorf(
			"data frame must be at least %d bytes long",
			DataFrameHeaderByteLength,
		)
	}

	frameKind := encoded[0]
	if frameKind != DataFrameKindData {
		return StreamDataFrame{}, fmt.Errorf("frameKind is not supported: %d", frameKind)
	}

	streamID := binary.BigEndian.Uint32(encoded[1:5])
	if err := validateStreamID(streamID); err != nil {
		return StreamDataFrame{}, err
	}

	payloadKind := encoded[5]
	if err := validatePayloadKind(payloadKind); err != nil {
		return StreamDataFrame{}, err
	}

	payload := make([]byte, len(encoded)-DataFrameHeaderByteLength)
	copy(payload, encoded[DataFrameHeaderByteLength:])

	return StreamDataFrame{
		FrameKind:   frameKind,
		StreamID:    streamID,
		PayloadKind: payloadKind,
		Payload:     payload,
	}, nil
}

File: packages/sandbox-session-protocol/src/data-frame.ts
export const DataFrameHeaderByteLength = 6;
export const DataFrameKindData = 0x01;
export const PayloadKindRawBytes = 0x01;
export const PayloadKindWebSocketText = 0x02;
export const PayloadKindWebSocketBinary = 0x03;
export const MaxStreamId = 0xffff_ffff;
export const DefaultStreamWindowBytes = 64 * 1024;
export const MaxStreamWindowBytes = DefaultStreamWindowBytes;

export interface StreamDataFrame {
  frameKind: number;
  streamId: number;
  payloadKind: number;
  payload: Uint8Array;
}

function validateStreamId(streamId: number): void {
  if (!Number.isInteger(streamId) || streamId <= 0 || streamId > MaxStreamId) {
    throw new Error(`streamId must be an integer between 1 and ${String(MaxStreamId)}`);
  }
}

function validatePayloadKind(payloadKind: number): void {
  if (
    payloadKind !== PayloadKindRawBytes &&
    payloadKind !== PayloadKindWebSocketText &&
    payloadKind !== PayloadKindWebSocketBinary
  ) {
    throw new Error(`payloadKind is not supported: ${String(payloadKind)}`);
  }
}

export function encodeDataFrame(input: {
  streamId: number;
  payloadKind: number;
  payload: Uint8Array;
}): Uint8Array {
  validateStreamId(input.streamId);
  validatePayloadKind(input.payloadKind);

  const encoded = new Uint8Array(DataFrameHeaderByteLength + input.payload.byteLength);
  const view = new DataView(encoded.buffer, encoded.byteOffset, encoded.byteLength);

  view.setUint8(0, DataFrameKindData);
  view.setUint32(1, input.streamId);
  view.setUint8(5, input.payloadKind);
  encoded.set(input.payload, DataFrameHeaderByteLength);

  return encoded;
}

export function decodeDataFrame(encoded: Uint8Array): StreamDataFrame {
  if (encoded.byteLength < DataFrameHeaderByteLength) {
    throw new Error(`data frame must be at least ${String(DataFrameHeaderByteLength)} bytes long`);
  }

  const view = new DataView(encoded.buffer, encoded.byteOffset, encoded.byteLength);
  const frameKind = view.getUint8(0);
  if (frameKind !== DataFrameKindData) {
    throw new Error(`frameKind is not supported: ${String(frameKind)}`);
  }

  const streamId = view.getUint32(1);
  validateStreamId(streamId);

  const payloadKind = view.getUint8(5);
  validatePayloadKind(payloadKind);

  return {
    frameKind,
    streamId,
    payloadKind,
    payload: encoded.slice(DataFrameHeaderByteLength),
  };
}
